### ISSUE TITLE
Add some developper helper in Plugin Class

### ISSUE THREAD

**Issue by Nico0084** [domogik/domogik#347]
Hi, 
In feedback from several developing plugin, particulary for non-xpl, it could be interresting to add some developper helper function in Plugin class to handle plugins devices.
I propose that beaucoup I write same function in different plugins, so I think they can be add in base class.
## Handling plugin devices uptodate :
- **init Plugin Class** : => add parameter _cb_update_devices = None_ to redirect a updating device from MQ

```
self._cb_update_devices  = None
```
- **def register_cb_update_devices(cb_update_devices)** => developper inherit Plugin call can register it's own updating devices process

```
self._cb_update_devices  = cb_update_devices 
```
- **def on_message(self, msgid, content)** =>  extend code by adding call to updateDevices
  and if developper overwrite it, specify to call **parent** at first (Plugin.on_message(self, msgid, content)).

```
        if msgid == "device.update":
            self._log.debug(u"New pub message {0}, {1}".format(msgid, content))
            self.threadingRefreshDevices()
```
- **def threadingRefreshDevices(self, max_attempt = 2)** => Run in thread get_device_list  to get free process.

```
        threading.Thread(None, self.refreshDevices, "th_refreshDevices", (), {"max_attempt": max_attempt`}).start()
```
- **def refreshDevices(self, max_attempt = 2)** => Call get_device_list from MQ, and call the callback if necessary.

```
        old_devices = self.devices
        devices = self.get_device_list(quit_if_no_device = False, max_attempt = max_attempt)
        if self._cb_update_devices is not None and  devices != old_devices :
             self._cb_update_devices(devices)
```
## Handling update  plugin devices global parameters:

```
def udpate_device_param(self, paramId, key=None, value=None):
        """Call DBHelper to update static device parameter"""
        with self._db.session_scope():
            config = self._db.udpate_device_param(paramId, key, value)
            self.log.debug(u"Setting global device parameter {0}, key {1} on value {2}".format(paramId, key, value))
            if config is not None:
                self.threadingRefreshDevices()
```
## Helper for DT_Types :
- **init Plugin Class** : => add parameter self._dataTypes = []
- Loading domogik DT_Types if necessary (at fist call of a data types request)

```
    def _loadDataType(self):
        mq_client  = MQSyncReq(self.zmq)
        msg = MQMessage()
        msg.set_action('datatype.get')
        result = mq_client.request('manager', msg.get(), timeout=10)
        if result :
            self._dataTypes = result.get_data()['datatypes']
            self.log.info(u"data_types list loaded.")
        else :
            self._log.warning(u"Error on retreive data_types list from MQ.")
```
- Getting DT_Type definition by name:
  
  ```
  def getDataType(self, name):
        """Return Datatype dict corresponding to name """
        if self._dataTypes == [] : self._loadDataType()
        for dT in self._dataTypes :
            if dT == name : return self._dataTypes[dT]
        return {}
  ```
## Helpers for detecting devices:
- Getting a product definition by Id:

```
    def get_product_by_id(self, productId):
        """Return the product(s) set in json corresponding to productId.
           productId must contain id."""
        products = []
        productId = productId.lower()
        for product in self.json_data['products']:
            if productId.find(product['id'].lower()) != -1:
                product[u'picture'] = self.get_picture_product(product)
                products.append(product)
        return products
```

```
    def get_picture_product(self, product):
        """Return product picture file from json plugin, else None.
           @param : product : the json dict definition from json plugin.
        """
        for ext in PRODUCTS_PICTURES_EXTENSIONS:
            file = "{0}.{1}".format(product['id'], ext)
            if os.path.isfile("{0}/{1}".format(self.get_products_directory(), file)):
                return file
        return None
```
- Getting sensor Plugin definition :
  
  ```
  def get_sensors_by_name(self, name):
        """Return the sensor(s) set in json corresponding to name in lower case"""
        sensors = {}
        name = name.lower()
        for sensor in self.json_data['sensors']:
            if self.json_data['sensors'][sensor]['name'].lower() == name :
                sensors[sensor] = self.json_data['sensors'][sensor]
        return sensors
  ```

```
  def get_sensor_by_id(self, id):
        """Return the sensor set in json corresponding to id"""
        for sensor in self.json_data['sensors']:
            if sensor == id:
                return self.json_data['sensors'][sensor]
        return {}
```
- Getting  command Plugin definition : 

```
    def get_commands_by_name(self, name):
        """Return the command(s) set in json corresponding to name in lower case"""
        cmds = {}
        name = name.lower()
        for cmd in self.json_data['commands']:
            for param in self.json_data['commands'][cmd]['parameters']:
                if param['key'].lower() == name :
                    cmds[cmd] = self.json_data['commands'][cmd]
        return cmds
```

```
    def get_command_by_id(self, id):
        """Return the command set in json corresponding to name in lower case"""
        for cmd in self.json_data['commands']:
            if cmd == id:
                return self.json_data['commands'][cmd]
        return {}
```

Probable there are others helpers we can add !
I can do the job to add these if you are agree :)

**Comment by Cereal2nd** [comment]
please create a pull request for this

only the def threadingRefreshDevices(self, max_attempt = 2) is not needed we should use the device.updated pub message for this

**Comment by Nico0084** [comment]
These implementations are done and tested with ozwave plugin, that works :)

All modifications are in my local branch "plugin-helper" (not a very good name, sorry ;)
what you want to do for reviewing code?
1 - push in my repository and you check in that place
2 - merge in my develop and PR to domogik 
3 - PR in a new domogik branch

There is documentation to write but I'm not sure of where to place in the development part

**Comment by Nico0084** [comment]
Merged in PR #350
